compiler/sem/semstmts.nim

```diff
@@ -1648,28 +1650,30 @@ proc semFor(c: PContext, n: PNode; flags: TExprFlags): PNode =
   if isCallExpr and call[0].kind == nkSym and
       call[0].sym.magic in {mFields, mFieldPairs}:
     result = semForFields(c, n, call[0].sym.magic)
-  elif isCallExpr and isClosureIterator(call[0].typ.skipTypes(abstractInst)):
-    # first class iterator:
-    result = semForVars(c, n, flags)
-  elif not isCallExpr or call[0].kind != nkSym or
-      call[0].sym.kind != skIterator:
-    if n.len == 3:
-      n[^2] = implicitIterator(c, "items", n[^2])
-    elif n.len == 4:
-      n[^2] = implicitIterator(c, "pairs", n[^2])
-    else:
-      localReport(c.config, n[^2], reportSem(rsemForExpectsIterator))
-    result = semForVars(c, n, flags)
   else:
+    if isCallExpr and isClosureIterator(call[0].typ.skipTypes(abstractInst)):
+      # first class iterator:
+      discard
+    elif not hasError and (not isCallExpr or call[0].kind != nkSym or
+        call[0].sym.kind != skIterator):
+      if n.len == 3:
+        n[^2] = implicitIterator(c, "items", n[^2])
+        hasError = n[^2].isError
+      elif n.len == 4:
+        n[^2] = implicitIterator(c, "pairs", n[^2])
+        hasError = n[^2].isError
+      else:
+        localReport(c.config, n[^2], reportSem(rsemForExpectsIterator))
     result = semForVars(c, n, flags)
-  if result.kind == nkError:
+  if hasError or result.kind == nkError:
```

Comment:
since `result` is the same as `n`, a single `var hasError = result[^2].kind == nkError` would work instead of changing each branch, this would allow for easy change of the `localReport` to an `nkError`, as well.

Reply:
see line 1657 `hasError` is used prevent `call[0]` cause `IndexDefect`

Reply:
Ah, good point.